### PR TITLE
Add ExistsFilter

### PR DESCRIPTION
**Pull request by teohhanhui**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | TODO
| Fixed tickets | N/A
| License       | MIT
| Doc PR        | TODO

IS NULL / IS NOT NULL (field / single-valued association)
or
IS EMPTY / IS NOT EMPTY (collection-valued association)

### CODE REVIEW

**Comment by vincentchalamon** [comment]
Please add unit & functional tests

**Comment by soyuka** [comment]
@teohhanhui I let you do the rebase thing okay?

**Comment by teohhanhui** [comment]
@vincentchalamon @soyuka I think we should allow the query parameter key ("exists") to be changed, right? Similar to "order".

**Comment by soyuka** [comment]
@teohhanhui I'm not sure that's needed but I can add it. Let me know.

**Comment by teohhanhui** [comment]
It has potential to conflict with a boolean property named "exists".

File: tests/Bridge/Doctrine/Orm/Filter/ExistsFilterTest.php [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+                'required' => false,
+            ],
+        ], $filter->getDescription($this->resourceClass));
+    }
+
+    /**
+     * Provides test data.
+     *
+     * Provides 4 parameters:
+     *  - order parameter name
```
Comment (teohhanhui) [inline tests/Bridge/Doctrine/Orm/Filter/ExistsFilterTest.php]: Outdated phpdoc.

File: tests/Fixtures/TestBundle/Entity/Dummy.php [hunk 2]
```diff
@@ -22,7 +22,7 @@
  *
  * @author Kévin Dunglas <dunglas@gmail.com>
  *
- * @ApiResource(attributes={"filters"={"my_dummy.search", "my_dummy.order", "my_dummy.date", "my_dummy.range", "my_dummy.boolean", "my_dummy.numeric"}})
+ * @ApiResource(attributes={"filters"={"my_dummy.search", "my_dummy.order", "my_dummy.date", "my_dummy.range", "my_dummy.boolean", "my_dummy.numeric", "my_dummy.exists"}})
```
Comment (teohhanhui) [inline tests/Fixtures/TestBundle/Entity/Dummy.php]: Can't we split this nicely into one line per filter (and sorted alphabetically)?

**Comment by soyuka** [comment]
Can you be more specific I didn't get that :laughing: 

Should I add a behat test for exists?

**Comment by teohhanhui** [comment]
We do have Behat tests for all the filters, e.g. https://github.com/api-platform/core/blob/master/features/doctrine/order_filter.feature

What I meant is, we can add a boolean "exists" property in the dummy resource, and enable the BooleanFilter for that property. It should not impact the functionality of the ExistsFilter, because it uses a different structure from the BooleanFilter for the query parameters.

**Comment by soyuka** [comment]
Okay adding that.

**Comment by vincentchalamon** [comment]
@teohhanhui 👍 for changing `exists` key. I also recommend a Behat test to check `does not exist` feature. /cc @soyuka

**Comment by soyuka** [comment]
@vincentchalamon doesn't a non-conflict test suffice? I understand the need to change the "order" key but for "exists" it seems overdone.

**Comment by soyuka** [comment]
did the modifications but I think travis has an issue when I do the push on your repository @teohhanhui

**Comment by teohhanhui** [comment]
I'd actually love to see `order_parameter_name` deprecated if we can find out / make sure that there's no conflict.

**Comment by soyuka** [comment]
@teohhanhui Sometimes you just want to name that "sort" though \o/

**Comment by vincentchalamon** [comment]
@soyuka 👍
But I still recommend a `does not exist` test 😉

**Comment by soyuka** [comment]
I've done that: https://github.com/api-platform/core/pull/906/files#diff-a6db6250a11c29a00faf3d4ae19a769fR10

File: features/doctrine/exists_filter.feature [hunk 3]
```diff
@@ -0,0 +1,61 @@
+Feature: Exists filter on collections
+  In order to retrieve large collections of resources
+  As a client software developer
+  I need to retrieve collections that exists
+
+  @dropSchema
+  Scenario: delete
+
+  @createSchema
+  Scenario: Get collection that does not exists
```
Comment (teohhanhui) [inline features/doctrine/exists_filter.feature]: Grammar: does not exist~s~

**Comment by teohhanhui** [comment]
One moment. Let me rebase...

File: features/doctrine/exists_filter.feature [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        "@context": {"pattern": "^/contexts/Dummy$"},
+        "@id": {"pattern": "^/dummies$"},
+        "@type": {"pattern": "^hydra:Collection$"},
+        "hydra:totalItems": {"type":"number", "maximum": 0},
+        "hydra:member": {
+          "type": "array",
+          "maxItems": 0
+        },
+        "hydra:view": {
+          "@id": {"pattern": "^/dummies\\?exists=[exists=0]"},
```
Comment (vincentchalamon) [inline features/doctrine/exists_filter.feature]: I think it should be `"^/dummies\\?exists[exists]=0"`, isn't it?

File: features/doctrine/exists_filter.feature [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        "@context": {"pattern": "^/contexts/Dummy$"},
+        "@id": {"pattern": "^/dummies$"},
+        "@type": {"pattern": "^hydra:Collection$"},
+        "hydra:totalItems": {"type":"number", "minimum": 3},
+        "hydra:member": {
+          "type": "array",
+          "minItems": 3
+        },
+        "hydra:view": {
+          "@id": {"pattern": "^/dummies\\?exists=[exists=1]"},
```
Comment (vincentchalamon) [inline features/doctrine/exists_filter.feature]: Should be `"^/dummies\\?exists[exists]=1"` I guess

**Comment by teohhanhui** [comment]
Rebased.

File: tests/Fixtures/TestBundle/Entity/Dummy.php [hunk 6]
```diff
@@ -130,6 +140,13 @@ class Dummy
      */
     public $nameConverted;
 
+    /**
+     * @var bool
+     *
+     * @ORM\Column(type="boolean", nullable=true, name="existsd")
```
Comment (vincentchalamon) [inline tests/Fixtures/TestBundle/Entity/Dummy.php]: Typo on name? `existsd` => `exists`

Comment (teohhanhui) [inline tests/Fixtures/TestBundle/Entity/Dummy.php]: EXISTS is an SQL keyword. But we should probably use a name that doesn't look like a typo.

`it_exists`? :rofl:

Comment (vincentchalamon) [inline tests/Fixtures/TestBundle/Entity/Dummy.php]: `is_exist` should be more consistent, no?

Comment (soyuka) [inline tests/Fixtures/TestBundle/Entity/Dummy.php]: @teohhanhui you do or I do the changes?

**Comment by teohhanhui** [comment]
The test failures indicate some of the other filters are not as strict as they should be...

File: tests/Fixtures/TestBundle/Entity/Dummy.php (hunk 6 above)
Comment (teohhanhui) [inline tests/Fixtures/TestBundle/Entity/Dummy.php]: @soyuka Do you think we should get another PR merged into 2.0 branch to make some of the other filters more strict in terms of which properties they can be applied to? I'd argue that it should not be considered a BC break, for example if the RangeFilter stopped working for boolean properties (lol) or if the SearchFilter stopped working with a numeric property (really really bad idea in the first place).

**Comment by soyuka** [comment]
It's just because I added that `exists` property no?

File: tests/Fixtures/TestBundle/Entity/Dummy.php (hunk 6 above)
Comment (soyuka) [inline tests/Fixtures/TestBundle/Entity/Dummy.php]: Hmm, something that would prevent from doing a NumericFilter on boolean for example?

I think that would be great indeed, with nice error messages to guide the developer.

**Comment by teohhanhui** [comment]
> It's just because I added that exists property no?

Yeah, and because of that we have to add the property in the tests for some other filters where `$properties` is `null` :disappointed:

**Comment by soyuka** [comment]
Yeah anyway I've already fixed the tests, pushing in a few minutes

**Comment by soyuka** [comment]
Though here we could not add a SearchFilter on `exists`. What can we do about that?

**Comment by teohhanhui** [comment]
@soyuka:
> Though here we could not add a SearchFilter on exists. What can we do about that?

What do you mean?

**Comment by soyuka** [comment]
Ie url: `localhost/api/exists=1` if `SearchFilter` and `ExistsFilter`  both exists, which one does take precedence? Will there be a conflict between those?

**Comment by teohhanhui** [comment]
Looks like we have no choice but to make the query parameter key
configurable then...

**Comment by soyuka** [comment]
Yes okay, I'll add it.

**Comment by soyuka** [comment]
Finally all green, sorry @Simperfit for the 0.06%

File: features/doctrine/date_filter.feature [hunk 7]
```diff
@@ -386,204 +386,25 @@ Feature: Date filter on collections
...
+        "@context": {"pattern": "^/contexts/Dummy$"},
+        "@id": {"pattern": "^/dummies$"},
+        "@type": {"pattern": "^hydra:Collection$"},
+        "hydra:totalItems": {"type":"number", "maximum": 0},
+        "hydra:member": {
+          "type": "array",
+          "maxItems": 0
+        },
+        "hydra:view": {
+          "@id": {"pattern": "^/dummies\\?exists=[exists]=0"},
```
Comment (vincentchalamon) [inline features/doctrine/date_filter.feature]: `"^/dummies\\?exists[exists]=0"`

File: features/doctrine/exists_filter.feature [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        "@context": {"pattern": "^/contexts/Dummy$"},
+        "@id": {"pattern": "^/dummies$"},
+        "@type": {"pattern": "^hydra:Collection$"},
+        "hydra:totalItems": {"type":"number", "maximum": 0},
+        "hydra:member": {
+          "type": "array",
+          "maxItems": 0
+        },
+        "hydra:view": {
+          "@id": {"pattern": "^/dummies\\?exists=[exists]=0"},
```
Comment (vincentchalamon) [inline features/doctrine/exists_filter.feature]: `"^/dummies\\?exists[exists]=0"`

File: features/doctrine/exists_filter.feature [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        "@context": {"pattern": "^/contexts/Dummy$"},
+        "@id": {"pattern": "^/dummies$"},
+        "@type": {"pattern": "^hydra:Collection$"},
+        "hydra:totalItems": {"type":"number", "minimum": 3},
+        "hydra:member": {
+          "type": "array",
+          "minItems": 3
+        },
+        "hydra:view": {
+          "@id": {"pattern": "^/dummies\\?exists=[exists]=1"},
```
Comment (vincentchalamon) [inline features/doctrine/exists_filter.feature]: `"^/dummies\\?exists[exists]=1"`

File: tests/Fixtures/TestBundle/Entity/Dummy.php [hunk 10]
```diff
@@ -255,4 +272,24 @@ public function getDummy()
     {
         return $this->dummy;
     }
+
+    /**
+     * Get exists.
+     *
+     * @return exists
```
Comment (vincentchalamon) [inline tests/Fixtures/TestBundle/Entity/Dummy.php]: `@return bool|null`

File: tests/Fixtures/TestBundle/Entity/Dummy.php [hunk 11]
```diff
@@ -255,4 +272,24 @@ public function getDummy()
...
+     */
+    public function getExists()
+    {
+        return $this->exists;
+    }
+
+    /**
+     * Set exists.
+     *
+     * @param exists the value to set
```
Comment (vincentchalamon) [inline tests/Fixtures/TestBundle/Entity/Dummy.php]: `@param bool|null $exists`

File: tests/Fixtures/TestBundle/Entity/Dummy.php [hunk 12]
```diff
@@ -255,4 +272,24 @@ public function getDummy()
...
+     */
+    public function setExists($exists)
```
Comment (vincentchalamon) [inline tests/Fixtures/TestBundle/Entity/Dummy.php]:
```php
public function setExists(bool $exists = null)
```

**Comment by soyuka** [comment]
I really have no clue why it fails on 7.1 deps=low...

**Comment by dunglas** [comment]
Can you rebase? It should fix the tests.

**Comment by teohhanhui** [comment]
@soyuka I think we can remove the `exists` property that we added to the dummy resource, and the related tests.

**Comment by soyuka** [comment]
Why? Tests do pass no? https://travis-ci.org/api-platform/core/builds/203123363

**Comment by teohhanhui** [comment]
Yes, they do pass. But they add no value, because what we were previously
aiming for is no longer necessary to be tested. :smile:

**Comment by soyuka** [comment]
Bump removed the new property, LGTM.

**Comment by teohhanhui** [comment]
@soyuka What do you think about https://github.com/api-platform/core/issues/70#issuecomment-281374347?

It should remove the need for customizing the query parameter key.

**Comment by soyuka** [comment]
You mean by using `?property[null]=true/false`? As you wish!

**Comment by teohhanhui** [comment]
`?property[exists]` instead of `?exists[property]`

**Comment by soyuka** [comment]
Okay, will be done tomorrow!

File: src/Bridge/Doctrine/Orm/Filter/ExistsFilter.php [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+        return $description;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function filterProperty(string $property, $value, QueryBuilder $queryBuilder, QueryNameGeneratorInterface $queryNameGenerator, string $resourceClass, string $operationName = null)
+    {
+        if (
+            !isset($value['exists']) ||
```
Comment (teohhanhui) [inline src/Bridge/Doctrine/Orm/Filter/ExistsFilter.php]: We should use a constant.

```php
const QUERY_PARAMETER_KEY = 'exists';
```

Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Filter/ExistsFilter.php]: done

File: tests/Fixtures/TestBundle/Entity/Dummy.php [hunk 14]
```diff
@@ -22,7 +22,17 @@
...
- * @ApiResource(attributes={"filters"={"my_dummy.search", "my_dummy.order", "my_dummy.date", "my_dummy.range", "my_dummy.boolean", "my_dummy.numeric"}})
+ * @ApiResource(attributes={
+ *     "filters"={
+ *         "my_dummy.boolean",
+ *         "my_dummy.date",
+ *         "my_dummy.exists",
+ *         "my_dummy.order",
+ *         "my_dummy.numeric",
+ *         "my_dummy.range",
+ *         "my_dummy.search"
```
Comment (teohhanhui) [inline tests/Fixtures/TestBundle/Entity/Dummy.php]: Trailing commas are supported, so let's add them. Same for next line...

File: tests/Fixtures/TestBundle/Entity/Dummy.php [hunk 15]
```diff
@@ -22,7 +22,17 @@
  *
  * @author Kévin Dunglas <dunglas@gmail.com>
  *
- * @ApiResource(attributes={"filters"={"my_dummy.search", "my_dummy.order", "my_dummy.date", "my_dummy.range", "my_dummy.boolean", "my_dummy.numeric"}})
+ * @ApiResource(attributes={
+ *     "filters"={
+ *         "my_dummy.boolean",
+ *         "my_dummy.date",
+ *         "my_dummy.exists",
+ *         "my_dummy.order",
```
Comment (teohhanhui) [inline tests/Fixtures/TestBundle/Entity/Dummy.php]: **n**umeric, **o**rder

Wrong ordering.

File: features/doctrine/date_filter.feature [hunk 16]
```diff
@@ -386,204 +386,25 @@ Feature: Date filter on collections
     Then the response status code should be 200
     And the response should be in JSON
     And the header "Content-Type" should be equal to "application/ld+json; charset=utf-8"
-    And the JSON should be equal to:
+    And the JSON should be valid according to this schema:
```
Comment (teohhanhui) [inline features/doctrine/date_filter.feature]: Why was this file changed? Merge cock-up? :stuck_out_tongue:

Comment (teohhanhui) [inline features/doctrine/date_filter.feature]: If we're fixing this, it should be done in another PR, I think.

(And the `?exists[exists]` below is from before...)

File: tests/Fixtures/TestBundle/Entity/Dummy.php [hunk 17]
```diff
@@ -22,7 +22,17 @@
...
+ * @ApiResource(attributes={
+ *     "filters"={
+ *         "my_dummy.boolean",
+ *         "my_dummy.date",
+ *         "my_dummy.exists",
+ *         "my_dummy.numeric",
+ *         "my_dummy.order",
+ *         "my_dummy.range",
+ *         "my_dummy.search",
+ *     }
```
Comment (teohhanhui) [inline tests/Fixtures/TestBundle/Entity/Dummy.php]: Trailing comma here too...

**Comment by soyuka** [comment]
Why are tests broken now? -_-, I updated the branch and removed a unused namespace import but I can't find why Behat fails (it's find locally)

**Comment by teohhanhui** [comment]
@soyuka Behatch botched up. Lol... https://github.com/Behatch/contexts/releases/tag/v2.6

**Comment by soyuka** [comment]
Lol, deprecate or maybe broke instead?

**Comment by teohhanhui** [comment]
Opened https://github.com/Behatch/contexts/issues/200 for now.

Yeah, I'd love for us to not have to depend on that package. We'll probably have to write our own JsonContext / RestContext etc., not that hard I think.

**Comment by dunglas** [comment]
I would prefer to get rid of Behat completely and write a nice API testing package. But it will require to rewrite our whole test suite.

**Comment by soyuka** [comment]
> I would prefer to get rid of Behat completely and write a nice API testing package. But it will require to rewrite our whole test suite.

OMG I'm so +1 on this... I started writing some api tests with nodejs (request), but it's lacking a few things to write clean and easy-to-write tests.

I'm all in if you need help on writing the tests again, and if you want to brainstorm on how we get to test APIs better!

File: features/doctrine/exists_filter.feature [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+        "@context": {"pattern": "^/contexts/Dummy$"},
+        "@id": {"pattern": "^/dummies$"},
+        "@type": {"pattern": "^hydra:Collection$"},
+        "hydra:totalItems": {"type":"number", "maximum": 0},
+        "hydra:member": {
+          "type": "array",
+          "maxItems": 0
+        },
+        "hydra:view": {
+          "@id": {"pattern": "^/dummies\\?dummyBoolean[exists]=0"},
```
Comment (antograssiot) [inline features/doctrine/exists_filter.feature]: This json schema is not correctly constructed. You could put anything in the pattern, tests would still pass.
See #967
(Same goes bellow)

Comment (soyuka) [inline features/doctrine/exists_filter.feature]: Thanks!

**Comment by soyuka** [comment]
@teohhanhui lmk if this is good to you we can merge it.

**Comment by teohhanhui** [comment]
@soyuka I think we still need another approval. :rofl:

**Comment by soyuka** [comment]
As you wish :stuck_out_tongue:

File: src/Bridge/Doctrine/Orm/Filter/ExistsFilter.php [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+     * @return bool
+     */
+    protected function isNullableField(string $property, string $resourceClass): bool
+    {
+        $propertyParts = $this->splitPropertyParts($property);
+        $metadata = $this->getNestedMetadata($resourceClass, $propertyParts['associations']);
+
+        $field = $propertyParts['field'];
+
+        if ($metadata->hasAssociation($field)) {
```
Comment (Simperfit) [inline src/Bridge/Doctrine/Orm/Filter/ExistsFilter.php]: Can't we refacto that to avoid that complexity ?

**Comment by soyuka** [comment]
I'm now officially a git rebase master \o/ @Simperfit done what I could (removed 3 lines of code :laughing: ).

File: src/Bridge/Doctrine/Orm/Filter/ExistsFilter.php (hunk 19 above)
Comment (teohhanhui) [inline src/Bridge/Doctrine/Orm/Filter/ExistsFilter.php]: You mean to reduce the cyclomatic complexity?

Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Filter/ExistsFilter.php]: Look at my commit (made my changes in a separate commit), I refactored it a bit:

```diff

         if ($metadata->hasAssociation($field)) {
              if ($metadata->isSingleValuedAssociation($field)) {
 -                if ($metadata instanceof ClassMetadataInfo) {
 -                    $associationMapping = $metadata->getAssociationMapping($field);
 -
 -                    return $this->isAssociationNullable($associationMapping);
 -                } else {
 +                if (!($metadata instanceof ClassMetadataInfo)) {
                      return false;
                  }
 +
 +                $associationMapping = $metadata->getAssociationMapping($field);
 +
 +                return $this->isAssociationNullable($associationMapping);
              }
  
              return true;
          }
  
 -        if ($metadata->hasField($field)) {
 -            if ($metadata instanceof ClassMetadataInfo) {
 -                return $metadata->isNullable($field);
 -            }
 +        if ($metadata instanceof ClassMetadataInfo && $metadata->hasField($field)) {
 +            return $metadata->isNullable($field);
          }
```

**Comment by teohhanhui** [comment]
Thanks @soyuka! ^^

**Comment by aggrosoft** [comment]
Just tested this, but it does not read the assocation properties when autowiring - is there any reason for this? Works fine when defining them by using the configuration

**Comment by soyuka** [comment]
@aggrosoft  I'm not sure how autowiring should work with variable constructors but you can check how I make it work in [the PR that adds Filter annotations](https://github.com/api-platform/core/pull/1117/files).

**Comment by aggrosoft** [comment]
Shouldn't something like this make the autowiring work:

https://github.com/aggrosoft/core/commit/d29384ba4646e23b5f0a363ae65cc1e9895174c8

**Comment by teohhanhui** [comment]
@aggrosoft I think you've found a (very old, probably introduced in #242 :scream:) bug!

Sorry for not noticing earlier. :see_no_evil: :rofl:

(See the diff around here: https://github.com/api-platform/core/pull/242/files#diff-023fcb894b9b3ad337dee44f2ba0a53aR242 - coincidentally also line 242 :four_leaf_clover:)

**Comment by teohhanhui** [comment]
Hmm... But I think at that time `getFieldNames` must have included the association names too, otherwise I wouldn't have written the code for that?!
